sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetime.scala

```diff
@@ -0,0 +1,229 @@
+import org.apache.spark.unsafe.types.UTF8String
+
+case class DateFormat(left: Expression, right: Expression)
+  extends BinaryExpression with ExpectsInputTypes {
+
+  override def dataType: DataType = StringType
+
+  override def expectedChildTypes: Seq[DataType] = Seq(TimestampType, StringType)
+
+  override def foldable: Boolean = left.foldable && right.foldable
+
+  override def nullable: Boolean = true
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = left.eval(input)
+    if (valueLeft == null) {
+      null
+    } else {
+      val valueRight = right.eval(input)
+      if (valueRight == null) {
+        null
+      } else {
+        val sdf = new SimpleDateFormat(valueRight.asInstanceOf[UTF8String].toString)
+        left.dataType match {
+          case TimestampType =>
+            UTF8String.fromString(sdf.format(new Date(valueLeft.asInstanceOf[Long] / 10000)))
+          case DateType =>
+            UTF8String.fromString(sdf.format(DateTimeUtils.toJavaDate(valueLeft.asInstanceOf[Int])))
+          case StringType =>
+            UTF8String.fromString(
+              sdf.format(DateTimeUtils.stringToTime(valueLeft.asInstanceOf[UTF8String].toString)))
+        }
+      }
+    }
+  }
+
+  override def toString: String = s"DateFormat($left, $right)"
+
+  override protected def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val sdf = "java.text.SimpleDateFormat"
+    val utf8 = "org.apache.spark.unsafe.types.UTF8String"
+    val dtUtils = "org.apache.spark.sql.catalyst.util.DateTimeUtils"
+
+    val eval1 = left.gen(ctx)
+    val eval2 = right.gen(ctx)
+
+    val calc = left.dataType match {
+      case TimestampType =>
+        s"""$utf8.fromString(sdf.format(new java.sql.Date(${eval1.primitive} / 10000)));"""
+      case DateType =>
+        s"""$utf8.fromString(
+          sdf.format($dtUtils.toJavaDate(${eval1.primitive})));"""
+      case StringType =>
+        s"""
+           $utf8.fromString(sdf.format(new java.sql.Date($dtUtils.stringToTime(${eval1.primitive}.toString()).getTime())));
+         """
+    }
+
+    s"""
+      ${eval1.code}
+      boolean ${ev.isNull} = ${eval1.isNull};
+      ${ctx.javaType(dataType)} ${ev.primitive} = ${ctx.defaultValue(dataType)};
+      if (!${ev.isNull}) {
+        ${eval2.code}
+        if (!${eval2.isNull}) {
+          $sdf sdf = new $sdf(${eval2.primitive}.toString());
+          ${ev.primitive} = $calc
+        } else {
+          ${ev.isNull} = true;
+        }
+      }
+    """
+  }
+}
+
+case class Year(child: Expression) extends UnaryExpression with ExpectsInputTypes {
+
+  override def dataType: DataType = IntegerType
+
+  override def expectedChildTypes: Seq[DataType] = Seq(DateType, StringType, TimestampType)
+
+  override def foldable: Boolean = child.foldable
+
+  override def nullable: Boolean = true
+
+  override def eval(input: InternalRow): Any = {
+    DateFormat(child, Literal("y")).eval(input) match {
```

Comment:
Is it okay to call `DateFormat(child, Literal("y")).eval(input)` or is there a more elegant way.

Is there a way to call `genCode` of `DataFormat`?